types/sqs-consumer/index.d.ts

```diff
@@ -1,29 +1,33 @@
-// Type definitions for sqs-consumer 3.5
+// Type definitions for sqs-consumer 3.7
 // Project: https://github.com/BBC/sqs-consumer
-// Definitions by: Daniel Chao <http://dchao.co/>, Eric Byers <http://www.github.com/EricByers>
+// Definitions by: Daniel Chao <http://dchao.co/>
+//                 Eric Byers <http://github.com/EricByers>
+//                 Ezinwa Okpoechi <http://github.com/BrainMaestro>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
 /// <reference types="node" />
 import { SQS } from "aws-sdk";
 
-export type ConsumerDone = (error?: Error) => void;
+export type ConsumerDone = (error?: SQSError) => void;
 
 export interface Options {
-  queueUrl: string;
-  handleMessage(message: SQS.Message, done: ConsumerDone): any;
-  region?: string;
-  attributeNames?: string[];
-  messageAttributeNames?: string[];
-  batchSize?: number;
-  visibilityTimeout?: number;
-  waitTimeSeconds?: number;
-  authenticationErrorTimeout?: number;
-  sqs?: SQS;
+    queueUrl: string;
+    handleMessage(message: SQS.Message, done: ConsumerDone): any;
+    region?: string;
+    attributeNames?: string[];
+    messageAttributeNames?: string[];
+    batchSize?: number;
+    visibilityTimeout?: number;
+    waitTimeSeconds?: number;
+    authenticationErrorTimeout?: number;
+    sqs?: SQS;
 }
 
-export interface Consumer extends NodeJS.EventEmitter {
-  start(): void;
-  stop(): void;
-}
+export class SQSError extends Error {}
 
-export function create(opts: Options): Consumer;
+export class Consumer extends NodeJS.EventEmitter {
```

Comment:
This should be something like:

```ts
declare class Consumer { ... }
export = Consumer
```

Reply:
https://github.com/Microsoft/dtslint/
blob/master/docs/strict-export-declare-modifiers.md 

That's why I switched

Reply:
You can put the rest of the exports in a `namespace`. For example:

```ts
declare namespace SQSConsumer {
    export type ConsumerDone = (error?: SQSError) => void;
    
    export interface Options {
        queueUrl: string;
        handleMessage(message: SQS.Message, done: ConsumerDone): any;
        region?: string;
        attributeNames?: string[];
        messageAttributeNames?: string[];
        batchSize?: number;
        visibilityTimeout?: number;
        waitTimeSeconds?: number;
        authenticationErrorTimeout?: number;
        sqs?: SQS;
    }
    
    export interface SQSError extends Error {}
}

declare class Consumer extends NodeJS.EventEmitter {
    constructor(options: SQSConsumer.Options);
    start(): void;
    stop(): void;
    static create(options: SQSConsumer.Options): Consumer;
}

export = Consumer;
```

Reply:
But why though? That's still an unusual way to export a class. And it doesn't prevent the `declare` linting issue I mentioned

Reply:
It's because of differences in the way that CommonJS modules work and how ES6 modules work.

The way that you need to be calling this code looks something like this in ES5:

```js
var Consumer = require('sqs-consumer');
```

If your export looks like this:

```ts
export class Consumer { ... }
```

You'll end up importing it like so:

```ts
import { Consumer } from "sqs-consumer";
```

The compiled javascript looks like this:

```js
var Consumer = require('sqs-consumer').Consumer;
```

If you export default, your import in TS looks like this:

```ts
import Consumer from "sqs-consumer"
```

which compiles to:

```js
var Consumer = require('sqs-consumer').default;
```

Neither of these exports are correct, and I think the only way to fix this is to use `export = ...`

Reply:
To get around the linting error, you can create a `tslint.json` file that removes that linting check.
```json
{
  "extends": "dtslint/dt.json",
  "rules": {
    "strict-export-declare-modifiers": false
  }
}
```

Reply:
Okay that makes a lot of sense. How do I import the `SQSConsumer` namespace.

Reply:
Hm.. I think to get this to work, your namespace and your class have to share the same name. So you'd probably need to change the name of your class to `SQSConsumer`.

Reference:

https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-class-d-ts.html

Reply:
Thanks. I ended doing that, but I changed the namespace to `Consmer` since the class in the actual library is called `Consumer`.

Reply:
Cool. I think you forgot to push up your changes?

Reply:
No they're up. You might be looking at an outdated diff